<a href="https://colab.research.google.com/github/aruntakhur/SitareUniversity/blob/main/MinMax_Search_Lab.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lab Assignment: Tic Tac Toe with Minimax Search

## Objective
In this lab, you will:
- Implement Tic Tac Toe using Python.
- Write state functions (`S0`, `Player`, `Actions`, `Result`, `Terminal`, `Utility`).
- Implement the **Minimax algorithm**.
- Count the number of states evaluated.
- Compare theoretical vs actual states explored.

In [4]:
# -----------------------------
# Part A: State Representation
# -----------------------------

# Initial state (state): empty 3x3 board
S0 = [" " for _ in range(9)]


def print_board(state):
    """Print the board in a 3x3 format"""
    print(state[0]+"|"+state[1]+"|"+state[2])
    print("-+-+-")
    print(state[3]+"|"+state[4]+"|"+state[5])
    print("-+-+-")
    print(state[6]+"|"+state[7]+"|"+state[8])



In [ ]:
# -----------------------------
# TODO Part A: Implement State Functions
# -----------------------------

def Player(state):
    """Return which player has the next turn"""
    # TODO: Implement using counts of X and O
    if state.count("X") == 0:
        return "X"
    elif state.count("X") > state.count("O"):
        return "O"
    else:
        return "X"  


def Actions(state):
    """Return list of legal moves"""
    # TODO: Return all indexes that are still empty
    return [i for i, cell in enumerate(state) if cell == " "]


def Result(state, action):
    """Return new state after action is taken"""
    # TODO: Copy the state and apply the move
    state[action] = Player(state)
    return state


def Terminal(state):
    """Return True if the state is terminal (win or draw)"""
    # TODO: Check for win or if no moves left
    if state.count("X") + state.count("O") == 9:
        return True
    if state[0] == state[1] == state[2] == "X" or state[3] == state[4] == state[5] == "X" or state[6] == state[7] == state[8] == "X" or state[0] == state[3] == state[6] == "X" or state[1] == state[4] == state[7] == "X" or state[2] == state[5] == state[8] == "X" or state[0] == state[4] == state[8] == "X" or state[2] == state[4] == state[6] == "X":
        return True
    elif state[0] == state[1] == state[2] == "O" or state[3] == state[4] == state[5] == "O" or state[6] == state[7] == state[8] == "O" or state[0] == state[3] == state[6] == "O" or state[1] == state[4] == state[7] == "O" or state[2] == state[5] == state[8] == "O" or state[0] == state[4] == state[8] == "O" or state[2] == state[4] == state[6] == "O":
        return True
    else:
        return False
    


def Utility(state):
    """Return +1 if X wins, -1 if O wins, else 0"""
    
    # 1. Check for a win
    if (state[0] == state[1] == state[2] == "X" or
        state[3] == state[4] == state[5] == "X" or
        state[6] == state[7] == state[8] == "X" or
        state[0] == state[3] == state[6] == "X" or
        state[1] == state[4] == state[7] == "X" or
        state[2] == state[5] == state[8] == "X" or
        state[0] == state[4] == state[8] == "X" or
        state[2] == state[4] == state[6] == "X"):
        return 1
    elif (state[0] == state[1] == state[2] == "O" or
          state[3] == state[4] == state[5] == "O" or
          state[6] == state[7] == state[8] == "O" or
          state[0] == state[3] == state[6] == "O" or
          state[1] == state[4] == state[7] == "O" or
          state[2] == state[5] == state[8] == "O" or
          state[0] == state[4] == state[8] == "O" or
          state[2] == state[4] == state[6] == "O"):
        return -1
        
    # 2. Check for a draw (only if there's no winner)
    elif state.count("X") + state.count("O") == 9:
        return 0
        
    # 3. Otherwise, the game is still in progress.
    else:
        return 0

In [35]:
# -----------------------------
# Part B: Minimax Implementation
# -----------------------------

def MAX_VALUE(state):
    global call_count
    call_count += 1

    if Terminal(state):
        return Utility(state)
        # return float("-inf")

    v = float("-inf")
    for action in Actions(state):
        # print(state,Player(state),action)
        res =  MIN_VALUE(Result(state.copy(),action))
        # print(f"THIS IS FROM MAX - {res}")
        v = max(v,res)
    return v


def MIN_VALUE(state):
    global call_count
    call_count += 1

    if Terminal(state):
        return Utility(state)
        # return float("inf")

    v = float("inf")
    for action in Actions(state):
        # print(state,Player(state),action)
        res =  MAX_VALUE(Result(state.copy(),action))
        # print(f"THIS IS FROM MIN - {res}")
        v = min(v,res)
    return v


def minimax_decision(state):
    """Return best move for current player"""
    global call_count
    call_count = 0

    player = Player(state)
    best_move = None

    if player == "X":
        best_val = float("-inf")
        for action in Actions(state):
            val = MAX_VALUE(Result(state.copy(), action))
            if val > best_val:
                best_val = val
                best_move = action
    else:
        best_val = float("inf")
        for action in Actions(state):
            # print(Actions(state))
            val = MIN_VALUE(Result(state.copy(), action))
            # print(f"{action} - {val}")
            if val < best_val:
                best_val = val
                best_move = action

    return best_move,call_count

In [ ]:
# -----------------------------
# Part C: Human vs AI Game
# -----------------------------
# Add Code to detect invalid move also
def valid_move (state,move):
    if move in Actions(state):
        return True
    return False

def play():
    state = S0.copy()

    while not Terminal(state):
        print_board(state)
        if Player(state) == "X":  # Human move
            move = int(input("Enter your move (1-9): ")) - 1
            while not valid_move(state,move):
                print("Position which you entered is taken by AI....")
                move = int(input("Enter your move (1-9): ")) - 1

        else:  # AI move
            print("AI is thinking...")
            move,call_count = minimax_decision(state.copy())
            print(f"AI explored {call_count} for making this decision...")

        state = Result(state, move)

    print_board(state)
    if Utility(state) == 1:
        print("X wins!")
    elif Utility(state) == -1:
        print("O wins!")
    else:
        print("It's a draw!")

In [37]:
play()

 | | 
-+-+-
 | | 
-+-+-
 | | 
 | | 
-+-+-
 |X| 
-+-+-
 | | 
AI is thinking...
AI explored 55504 for making this decision...
O| | 
-+-+-
 |X| 
-+-+-
 | | 
O| |X
-+-+-
 |X| 
-+-+-
 | | 
AI is thinking...
AI explored 932 for making this decision...
O| |X
-+-+-
 |X| 
-+-+-
O| | 


ValueError: invalid literal for int() with base 10: ''

# -----------------------------
# Part D: Experiment and Analysis
# -----------------------------

1. Run the game where X plays first in the center (position 5).
2. Note down the number of states AI explored.
3. Compare with:
   - 8! = 40,320 (leaf-only sequences)
   - 8+8⋅7+8⋅7⋅6+⋯+8! = 109,600 (all partial states)
4. Explain why actual count (~55,504) is smaller than 109,600.


# -----------------------------
# Part E: Written Questions
# -----------------------------

Answer briefly:

1. Why does Minimax need to evaluate so many states in Tic Tac Toe?
2. Difference between full tree size and actual explored states?
3. How do early wins reduce the number of states?
4. What challenges would arise if applying Minimax to Chess?
